pyomo/util/model_size.py

```diff
@@ -0,0 +1,155 @@
+import logging
+
+from pyomo.core import Block, Constraint, Var
+from pyomo.core.expr import current as EXPR
+from pyomo.core.kernel import ComponentSet
+from pyomo.gdp import Disjunct, Disjunction
+from pyutilib.misc import Container
+
+
+default_logger = logging.getLogger('pyomo.util.model_size')
+default_logger.setLevel(logging.INFO)
+
+
+class ModelSizeReport(Container):
+    """Stores model size information.
+
+    Active blocks are those who have an active flag of True and whose parent,
+    if exists, is an active block or an active Disjunct.
+
+    Active constraints are those with an active flag of True and: are reachable
+    via an active Block, are on an active Disjunct, or are on a disjunct with
+    indicator_var fixed to 1 with active flag True.
+
+    Active variables refer to the presence of the variable on an active
+    constraint, or that the variable is an indicator_var for an active
+    Disjunct.
+
+    Active disjuncts refer to disjuncts with an active flag of True, have an
+    unfixed indicator_var, and who participate in an active Disjunction.
+
+    Active disjunctions follow the same rules as active constraints.
+
+    """
+    pass
+
+
+def build_model_size_report(model):
+    """Build a model size report object."""
+    report = ModelSizeReport()
+    active_disjunctions = ComponentSet()
+    active_disjuncts = ComponentSet()
+    fixed_true_disjuncts = ComponentSet()
+    active_constraints = ComponentSet()
+    active_vars = ComponentSet()
+    new_containers = (model,)
+
+    while new_containers:
+        new_active_disjunctions = ComponentSet()
+        new_active_disjuncts = ComponentSet()
+        new_fixed_true_disjuncts = ComponentSet()
+        new_active_constraints = ComponentSet()
+
+        for container in new_containers:
+            (next_active_disjunctions,
+             next_fixed_true_disjuncts,
+             next_active_disjuncts,
+             next_active_constraints) = _process_active_container(container)
+            new_active_disjunctions.update(next_active_disjunctions)
+            new_active_disjuncts.update(next_active_disjuncts)
+            new_fixed_true_disjuncts.update(next_fixed_true_disjuncts)
+            new_active_constraints.update(next_active_constraints)
+
+        new_containers = ((new_active_disjuncts - active_disjuncts) |
+                          (new_fixed_true_disjuncts - fixed_true_disjuncts))
+
+        active_disjunctions.update(new_active_disjunctions)
+        active_disjuncts.update(new_active_disjuncts)
+        fixed_true_disjuncts.update(new_fixed_true_disjuncts)
+        active_constraints.update(new_active_constraints)
+
+    active_vars.update(
+        var for constr in new_active_constraints
+        for var in EXPR.identify_variables(
+            constr.body, include_fixed=False))
+    active_vars.update(
+        disj.indicator_var for disj in active_disjuncts)
+
+    report.active = Container()
```

Comment:
I worry about using "active" because of the potential of confusion between "active constraints" (where value(body) is at a bound) and "activated constraints" (i.e., not deactivated).  What do you think of renaming this "activated"?

Reply:
That makes sense.